File: 1-100/131-palindrome-partitioning.cpp
vector<string> list;
vector<vector<string>> ans;
vector<vector<string>> partition(string s) {
    backtracking(0, s);
    return ans;
}

void backtracking(int start, string &s) {
    if (start == s.size()) {
        ans.push_back(list);
        return;
    }
    for (int i = start; i < s.size(); ++i) {
        if (isPalindrome(s, start, i)) {
            list.push_back(s.substr(start, i-start+1));
            backtracking(i+1, s);
            list.pop_back();
        }
    }
}

bool isPalindrome(string &s, int l, int r) {
    while (l < r) {
        if (s[l++] != s[r--]) 
            return false;
    }
    return true;
}

File: python version/131-palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        n = len(s)

        ans = []
        curr = []

        def backtracking(start):
            if start == n:
                ans.append(curr[:])
                return
            for end in range(start+1, n+1):
                substr = s[start:end]
                if is_valid(substr):
                    curr.append(substr)
                    backtracking(end)
                    curr.pop()
        
        def is_valid(substr):
            left = 0
            right = len(substr)-1
            while left < right:
                if substr[left] != substr[right]:
                    return False
                else:
                    left += 1
                    right -= 1
            return True

        backtracking(0)

        return ans